src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,367 @@
+        key (get entity pk)]
+    [(:entity_id entity)
+     (serdes.hash/identity-hash (db/select-one model pk key)) ; TODO This sucks for identity-hash!
+     key]))
+
+(defn- ingested-type
+  "The dispatch function for several of the load multimethods: dispatching on the type of the incoming entity."
+  [ingested]
+  (-> ingested :serdes/meta :type))
+
+(defmulti serdes-dependencies
+  "Given an entity map as ingested (not a Toucan entity) returns a (possibly empty) list of its dependencies, where each
+  dependency is represented by either the entity ID or identity hash of the target entity.
+
+  Keyed on the model name.
+  Default implementation returns an empty vector, so only models that have dependencies need to implement this."
+  ingested-type)
+
+(defmethod serdes-dependencies :default [_]
+  [])
+
+(defmulti load-xform
+  "Given the incoming vanilla map as ingested, transform it so it's suitable for sending to the database (in eg.
+  [[db/simple-insert!]]).
+  For example, this should convert any foreign keys back from a portable entity ID or identity hash into a numeric
+  database ID. This is the mirror of [[extract-one]], in spirit. (They're not strictly inverses - [[extract-one]] drops
+  the primary key but this need not put one back, for example.)
+
+  By default, this just calls [[load-xform-basics]].
+  If you override this, call [[load-xform-basics]] as well."
+  ingested-type)
+
+(defn load-xform-basics
+  "Performs the usual steps for an incoming entity:
+  - Drop :serdes/meta
+
+  You should call this as a first step from any implementation of [[load-xform]].
+
+  This is a mirror (but not precise inverse) of [[extract-one-basics]]."
+  [ingested]
+  (dissoc ingested :serdes/meta))
+
+(defmethod load-xform :default [ingested]
+  (load-xform-basics ingested))
+
+(defmulti load-update!
+  "Called by the default [[load-one!]] if there is a corresponding entity already in the appdb.
+  The first argument is the model name, the second the incoming map we're deserializing, and the third is the Toucan
+  entity found in the appdb.
+
+  Defaults to a straightforward [[db/update!]], and you may not need to update it.
+
+  Keyed on the model name (the first argument), because the second argument doesn't have its `:serdes/meta` anymore.
+
+  Returns the primary key of the updated entity."
+  (fn [model _ _] model))
+
+(defmethod load-update! :default [model-name ingested local]
+  (let [model (db/resolve-model (symbol model-name))
+        pk    (models/primary-key model)
+        id    (get local pk)
+        ; Get a WHERE clause, but then strip off the WHERE part to include it in the JDBC call below.
+        ;where (update (db/honeysql->sql {:where [:= pk id]}) 0
+        ;              #(.substring 5))
+        ]
+    (log/tracef "Upserting %s %d: old %s new %s" model-name id (pr-str local) (pr-str ingested))
+    ; Using the two-argument form of [[db/update!]] that takes the model and a HoneySQL form for the actual update.
```

Comment:
oh, I didn't know that the 2-arity skip the types and hooks, I wonder if this was intentional?

Either way I think it's worth a more detailed comment saying that this is bc we want to skip all toucan's magics